p2p/ipld/plugin/plugin.go

```diff
@@ -0,0 +1,349 @@
+	return nil
+}
+
+func (l LazyLedgerPlugin) RegisterInputEncParsers(iec coredag.InputEncParsers) error {
+	iec.AddParser("raw", DagParserFormatName, DataSquareRowOrColumnRawInputParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) Name() string {
+	return "LazyLedger"
+}
+
+func (l LazyLedgerPlugin) Version() string {
+	return "0.0.0"
+}
+
+func (l LazyLedgerPlugin) Init(env *plugin.Environment) error {
+	return nil
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares>
+//
+// To determine the share and the namespace size the constants
+// types.ShareSize and types.NamespaceSize are used.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]node.Node, error) {
+	const extendedSquareSize = 256
+	br := bufio.NewReader(r)
+	nodes := make([]node.Node, 0, extendedSquareSize)
+	nodeCollector := func(hash []byte, children ...[]byte) {
+		cid, err := cidFromNamespacedSha256(hash)
+		if err != nil {
+			panic(fmt.Sprintf("nmt lib created a malformed hash: %s", err))
+		}
+		isLeaf := len(children) == 1
+		if isLeaf {
+			prependNode(nmtLeafNode{
+				cid:  cid,
+				Data: children[0],
+			}, &nodes)
+		} else if len(children) == 2 {
+			prependNode(nmtNode{
+				cid: cid,
+				l:   children[0],
+				r:   children[1],
+			}, &nodes)
+		} else {
+			panic("expected a binary tree")
+		}
+	}
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(nodeCollector),
+	)
+	for {
+		share := make([]byte, shareSize+namespaceSize)
+		if _, err := io.ReadFull(br, share); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if err := n.Push(share[:namespaceSize], share[namespaceSize:]); err != nil {
+			return nil, err
+		}
+	}
+	// to trigger the collection of nodes:
+	_ = n.Root()
+	return nodes, nil
+}
+
+func prependNode(newNode node.Node, nodes *[]node.Node) {
+	*nodes = append(*nodes, node.Node(nil))
+	copy((*nodes)[1:], *nodes)
+	(*nodes)[0] = newNode
+}
+
+func NmtNodeParser(block blocks.Block) (node.Node, error) {
+	const (
+		// length of the domain separator for leaf and inner nodes:
+		prefixOffset = 1
+		// nmtHashSize = flagSize+sha256.Size
+		nmtHashSize = namespaceSize*2+sha256.Size
+	)
+	var (
+		leafPrefix  = []byte{nmt.LeafPrefix}
+		innerPrefix = []byte{nmt.NodePrefix}
+	)
+	data := block.RawData()
+	if len(data) == 0 {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: nil,
+		}, nil
+	}
+	domainSeparator := data[:prefixOffset]
+	if bytes.Equal(domainSeparator, leafPrefix) {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: data[prefixOffset:],
+		}, nil
+	} else if bytes.Equal(domainSeparator, innerPrefix) {
+		return nmtNode{
+			cid: block.Cid(),
+			l:   data[prefixOffset : prefixOffset+nmtHashSize],
+			r:   data[prefixOffset+nmtHashSize:],
+		}, nil
+	}
```

Comment:
```suggestion
	}
	if bytes.Equal(domainSeparator, innerPrefix) {
		return nmtNode{
			cid: block.Cid(),
			l:   data[prefixOffset : prefixOffset+nmtHashSize],
			r:   data[prefixOffset+nmtHashSize:],
		}, nil
	}
```
Another matter of opinion, I don't think the else statement should be kept. Since it's three options, a switch statement might also be the most readable.

Reply:
I agree, will commit as suggested.